services/twofa.go

```diff
@@ -0,0 +1,292 @@
+	if t.keyServiceClient != nil {
+		if err := t.makeKeyServiceValidateRequest(accountID, code); err != nil {
+			return err
+		}
+	} else {
+		if len(code) != 6 {
+			return util.ErrBadTOTPCode
+		}
+		// Otherwise, validate the code locally
+		secret, err := t.ds.GetTOTPKey(accountID)
+		if err != nil {
+			return err
+		}
+
+		// Standard validation with one period of time drift in either direction
+		opts := totp.ValidateOpts{
+			Skew:   1,
+			Digits: otp.DigitsSix,
+		}
+		valid, err := totp.ValidateCustom(code, secret, time.Now().UTC(), opts)
+		if err != nil {
+			return fmt.Errorf("failed to validate TOTP code: %w", err)
+		}
+		if !valid {
+			return util.ErrBadTOTPCode
+		}
+	}
+
+	if !t.isKeyService {
+		// Check if code was already used and store it if not
+		if err := t.ds.CheckAndStoreTOTPCodeUsed(accountID, code); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// makeKeyServiceValidateRequest sends a request to the key service to validate a TOTP code
+func (t *TwoFAService) makeKeyServiceValidateRequest(accountID uuid.UUID, code string) error {
+	type totpValidateRequest struct {
+		AccountID uuid.UUID `json:"accountId"`
+		Code      string    `json:"code"`
+	}
+
+	reqBody := totpValidateRequest{
+		AccountID: accountID,
+		Code:      code,
+	}
+
+	var response struct{}
+	return t.keyServiceClient.MakeRequest(http.MethodPost, "/v2/server_keys/totp/validate", reqBody, &response)
+}
+
+// DeleteTOTPKey deletes a TOTP key for an account
+func (t *TwoFAService) DeleteTOTPKey(accountID uuid.UUID) error {
+	if t.keyServiceClient != nil {
+		return t.makeKeyServiceDeleteRequest(accountID)
+	}
+
+	// Otherwise, delete the key locally
+	return t.ds.DeleteTOTPKey(accountID)
+}
+
+// makeKeyServiceDeleteRequest sends a request to the key service to delete a TOTP key
+func (t *TwoFAService) makeKeyServiceDeleteRequest(accountID uuid.UUID) error {
+	return t.keyServiceClient.MakeRequest(
+		http.MethodDelete,
+		fmt.Sprintf("/v2/server_keys/totp/%s", accountID.String()),
+		nil,
+		nil)
+}
+
+// GenerateTOTPQRCode generates a QR code image for a TOTP key and returns it as a base64 encoded PNG string
+func (t *TwoFAService) GenerateTOTPQRCode(key *otp.Key) (string, error) {
+	// Generate QR code image with the configured size
+	img, err := key.Image(t.qrSize, t.qrSize)
+	if err != nil {
+		return "", err
+	}
+
+	// Encode as PNG
+	var buf bytes.Buffer
+	if err := png.Encode(&buf, img); err != nil {
+		return "", err
+	}
+
+	// Convert to base64
+	encoded := base64.StdEncoding.EncodeToString(buf.Bytes())
+	return "data:image/png;base64," + encoded, nil
+}
+
+// GenerateAndStoreRecoveryKey generates a 32-character recovery key
+// and stores its hash in the database for the specified account
+func (t *TwoFAService) GenerateAndStoreRecoveryKey(accountID uuid.UUID) (string, error) {
+	const keyLength = 32
+	const randLength = keyLength*8/5 + 2 // 32 bytes * 8 bits/byte / 5 bits/char + 2 padding
```

Comment:
32 characters in base32 = 32 chars * 5 bits/char = 160 bits
160 bits / 8 bits / char = 20 bytes

```suggestion
	const randLength = keyLength*5/8 // 32 chars * 5 bits/char / 8 bits/byte = 20 bytes
```

Reply:
What are the 2 padding bytes for?

Reply:
woops! not sure what i was doing here.